File: server/src/core/Deck.ts
import { Card, Rank, Suit } from "../shared/types.js";

// ─── Interface ────────────────────────────────────────────────────────────────

export interface IDeck {
  draw(): Card;
}

// ─── Standard shuffled deck ───────────────────────────────────────────────────

const SUITS: Suit[] = ["♣", "♦", "♥", "♠"];
const RANKS: Rank[] = ["2","3","4","5","6","7","8","9","10","J","Q","K","A"];

export class Deck implements IDeck {
  private cards: Card[];
  private index = 0;

  constructor() {
    this.cards = [];
    for (const suit of SUITS) {
      for (const rank of RANKS) {
        this.cards.push({ suit, rank });
      }
    }
    this.shuffle();
  }

  private shuffle() {
    for (let i = this.cards.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]];
    }
  }

  public draw(): Card {
    if (this.index >= this.cards.length) {
      // Reshuffle when exhausted
      this.index = 0;
      this.shuffle();
    }
    return this.cards[this.index++];
  }
}

File: client/Assets/Scripts/Domain/Card/Deck.cs
using System;
using System.Collections.Generic;

public class Deck : IDeck
{
  private Stack<Card> _cards;
  private Random _rng;

  public Deck(int seed = 0)
  {
    _rng = seed == 0 ? new Random() : new Random(seed);
    Initialize();
    Shuffle();
  }

  private void Initialize()
  {
    var list = new List<Card>();

    foreach (Suit suit in Enum.GetValues(typeof(Suit)))
    {
      foreach (Rank rank in Enum.GetValues(typeof(Rank)))
      {
        list.Add(new Card { Suit = suit, Rank = rank });
      }
    }

    _cards = new Stack<Card>(list);
  }

  public void Shuffle()
  {
    var list = new List<Card>(_cards);
    int n = list.Count;

    while (n > 1)
    {
      n--;
      int k = _rng.Next(n + 1);
      (list[k], list[n]) = (list[n], list[k]);
    }

    _cards = new Stack<Card>(list);
  }

  public Card Draw()
  {
    if (_cards.Count == 0)
      throw new InvalidOperationException("Deck empty");

    return _cards.Pop();
  }
}
